cloud_storage_clients/s3_client: Logging Upgrade

joe-redpanda opened the pull request:
s3_client discards the body of error responses if theyre not xml formatted. This is a problem because some s3 tests periodically fail on Bad Request with no further description of what failed the request.

This pr adds text logging for non-xml errors.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes
### Improvements
* Allows s3 client to log non-xml error response bodies

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=26589) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=26589) before we can accept your contribution.<br/><hr/>**joe-redpanda** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/redpanda-data/redpanda?pullRequest=26589) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=26589) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=26589) before we can accept your contribution.<br/><hr/>**joe-redpanda** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/redpanda-data/redpanda?pullRequest=26589) it.</sub>

dotnwat commented on src/v/cloud_storage_clients/s3_client.cc:
```diff
@@ -428,6 +430,23 @@ rest_error_response parse_xml_rest_error_response(iobuf&& buf) {
 }
 
+ss::future<rest_error_response> parse_unknown_format_error_response(
+  boost::beast::http::status status, iobuf&& buf) {
```
take iobuf by value. it's very cheap to move it and then we don't have references in coroutine parameters

dotnwat commented on src/v/cloud_storage_clients/s3_client.cc:
```diff
@@ -428,6 +430,23 @@ rest_error_response parse_xml_rest_error_response(iobuf&& buf) {
+      buf.size_bytes());
+    auto input_stream = make_iobuf_input_stream(std::move(buf));
+    auto string_buffer = co_await ss::util::read_entire_stream_contiguous(
+      input_stream);
```
how about factoring this out along with the same code in src/v/cloud_storage/tests/anomalies_detector_test.cc into a generic helper?

dotnwat commented on src/v/cloud_storage_clients/s3_client.cc:
```diff
@@ -428,6 +430,23 @@ rest_error_response parse_xml_rest_error_response(iobuf&& buf) {
+      "unknown s3 rest error format of size: {}",
+      buf.size_bytes());
+    auto input_stream = make_iobuf_input_stream(std::move(buf));
+    auto string_buffer = co_await ss::util::read_entire_stream_contiguous(
```
how big might the string be? linearizing a large chunk of text could be problematic for our allocation limits?

joe-redpanda replied:
Yeah thats a good question. AFAIK small. I tried to think of a reasonable limit s.t. we don't dump infinite amounts of text into the log. let me set a limit at 4k or so and just log if its incomplete

dotnwat commented on src/v/cloud_storage_clients/s3_error.cc:
```diff
@@ -625,4 +625,10 @@ std::string_view rest_error_response::resource() const noexcept {
 }
 
+std::ostream& operator<<(std::ostream& o, const rest_error_response err) {
+    return o << "code: " << err._code_str << " _message: " << err._message
+             << " request_id: " << err._request_id
+             << " resource: " << err._resource;
```
use fmt for formatting

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--5436fdd5-811e-411d-ab41-e6e11f5d7ea6-->
<!--TEST RESULTS-->
<!--c641d90b-4fd2-4c66-b79a-036bba373a44-->
<!--TEST RESULTS-->
<!--b46b1aef-742a-43b1-a093-a6bcbc5ea152-->
<!--TEST RESULTS-->
<!--bac5ec0e-ea37-4f2e-8545-bf114225c05e-->
<!--TEST RESULTS-->
<!--cde41229-246f-4b2d-bbc9-7d6e1cef9ffa-->
<!--TEST RESULTS-->
<!--ab2a9448-dc05-40b6-86fd-db73e46394a4-->
<!--TEST RESULTS-->
<!--a9d3466b-1216-43f6-bdfe-465a4efb2326-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67929">build#67929</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-------------------------|:---------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| IcebergUsageTest         | test_iceberg_usage   | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "spark"}                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67929#0197a967-8ddd-4cb4-b77f-d502d292137b | FLAKY         | 14/21    | upstream reliability is '86.49468892261002'. current run reliability is '66.66666666666666'. drift is 19.82802 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67929#0197a966-5277-4183-beeb-cb73c846c452 | FLAKY         | 20/21    | upstream reliability is '99.42857142857143'. current run reliability is '95.23809523809523'. drift is 4.19048 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": true}          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67929#0197a966-5277-4183-beeb-cb73c846c452 | FLAKY         | 20/21    | upstream reliability is '98.96907216494846'. current run reliability is '95.23809523809523'. drift is 3.73098 and the allowed drift is set to 50. The test should PASS  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68230">build#68230</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-------------------------|:---------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68230#0197c3ee-499b-4f06-bdc9-4c0b781e5458 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS              |
| DisablingPartitionsTest  | test_disable         |                                                                                                                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68230#0197c3ee-e4f2-4356-9d9b-04d9cd75f4e9 | FLAKY         | 14/21    | upstream reliability is '91.13924050632912'. current run reliability is '66.66666666666666'. drift is 24.47257 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68272">build#68272</a></summary>


| test_class                | test_method                             | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:----------------------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ArchivalTest              | test_connection_flicker                 | {"cloud_storage_type": 2} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68272#0197c71e-73e4-4c7e-9f26-52e9592705e9 | FLAKY         | 19/21    | upstream reliability is '98.98989898989899'. current run reliability is '90.47619047619048'. drift is 8.51371 and the allowed drift is set to 50. The test should PASS |
| RecreateTopicMetadataTest | test_recreated_topic_metadata_are_valid | {"replication_factor": 3} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68272#0197c71e-73e8-4dde-b2c8-f77dc9ddcb52 | FLAKY         | 20/21    | upstream reliability is '97.83889980353635'. current run reliability is '95.23809523809523'. drift is 2.6008 and the allowed drift is set to 50. The test should PASS  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68439">build#68439</a></summary>


| test_class                   | test_method                | test_arguments                       | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:-----------------------------|:---------------------------|:-------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageTimingStressTest | test_cloud_storage         | {"cleanup_policy": "compact,delete"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68439#0197d1ab-6822-4cf2-957d-ee866c3ed799 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS            |
| ClusterConfigTest            | test_valid_settings        |                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68439#0197d1ac-469d-4bf8-a70d-878adc27ebb1 | FLAKY         | 17/21    | upstream reliability is '83.6231884057971'. current run reliability is '80.95238095238095'. drift is 2.67081 and the allowed drift is set to 50. The test should PASS |
| DataMigrationsApiTest        | test_mount_inexistent      |                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68439#0197d1ac-469a-4d28-b434-8880ce189292 | FLAKY         | 20/21    | upstream reliability is '99.43019943019942'. current run reliability is '95.23809523809523'. drift is 4.1921 and the allowed drift is set to 50. The test should PASS |
| TopicDeleteCloudStorageTest  | drop_lifecycle_marker_test | {"cloud_storage_type": 2}            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68439#0197d1ac-4698-4c3a-9ce2-1a64c88792a0 | FLAKY         | 20/21    | upstream reliability is '97.2386587771203'. current run reliability is '95.23809523809523'. drift is 2.00056 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68465">build#68465</a></summary>


| test_class                | test_method                    | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:-------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MaintenanceTest           | test_maintenance_sticky        | {"use_rpk": false}                                                                                                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68465#0197d2e8-4185-4eb0-b495-aa361ebd8a98 | FLAKY         | 20/21    | upstream reliability is '97.35772357723577'. current run reliability is '95.23809523809523'. drift is 2.11963 and the allowed drift is set to 50. The test should PASS |
| PartitionMoveInterruption | test_cancelling_partition_move | {"compacted": false, "force_back": false, "recovery": "no_recovery", "replication_factor": 3, "unclean_abort": false}                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68465#0197d2e9-e39a-4c87-924b-a144b4f6137a | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| RaftAvailabilityTest      | test_controller_node_isolation |                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68465#0197d2e8-4180-4dd9-81c8-a5c56218ca21 | FLAKY         | 19/21    | upstream reliability is '97.13024282560706'. current run reliability is '90.47619047619048'. drift is 6.65405 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest  | test_node_operations           | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68465#0197d2e8-4184-4501-8658-dc93ce5bc66c | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS             |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68596">build#68596</a></summary>


| test_class            | test_method                               | test_arguments                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:----------------------|:------------------------------------------|:--------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodePoolMigrationTest | test_migrating_redpanda_nodes_to_new_pool | {"balancing_mode": "off", "cleanup_policy": "compact,delete", "test_mode": "tiered_storage_fast_moves"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68596#0197eaff-0c08-4010-b2d5-1f6ae72aee10 | FLAKY         | 29/30    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
| NodePoolMigrationTest | test_migrating_redpanda_nodes_to_new_pool | {"balancing_mode": "off", "cleanup_policy": "compact,delete", "test_mode": "tiered_storage_fast_moves"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68596#0197eaff-0c09-4f65-9a46-a9ae8813811b | FLAKY         | 29/30    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68763">build#68763</a></summary>


| test_class                   | test_method          | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:-----------------------------|:---------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterConfigTest            | test_valid_settings  |                                                                                                                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68763#0197f5b1-b90c-420a-8b86-4670e5169d35 | FLAKY         | 17/21    | upstream reliability is '83.11808118081181'. current run reliability is '80.95238095238095'. drift is 2.1657 and the allowed drift is set to 50. The test should PASS |
| FeaturesMultiNodeUpgradeTest | test_rollback        |                                                                                                                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68763#0197f5b1-b907-4caf-992d-783b83be80c6 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS             |
| RandomNodeOperationsTest     | test_node_operations | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68763#0197f5b3-0d1a-42fa-81d4-aa4f980231a1 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS             |
</details>
<!--/body-->

dotnwat commented on src/v/bytes/util.cc:
```diff
@@ -0,0 +1,19 @@
+
+ss::future<ss::sstring>
+iobuf_to_string(iobuf buf, std::optional<size_t> size_limit) {
+    auto input_stream = make_iobuf_input_stream(std::move(buf));
+    if (size_limit) {
+        auto var = co_await input_stream.read_up_to(size_limit.value());
+        co_return ss::sstring(var.begin(), var.end());
+    }
+    co_return co_await ss::util::read_entire_stream_contiguous(input_stream);
```
what about this which already exists

```
iobuf buf;
const_iobuf_parser p(buf);
auto s = p.read_string();
```

i think all we'd need to do is add some ability to have a max size or whatever.

dotnwat replied:
the we wouldn't need an enitre new util.cc thing

joe-redpanda replied:
iobuf_parser::to_string validates utf8

which evidently anomaly detector does not obey as it responds with an sstring dump of what seems to be a serde message

Given that s3 only replies with utf encoded data, i think the answer is to drop the alignment with anomaly detector and just have the s3 client use iobuf_parser

dotnwat replied:
>iobuf_parser::to_string validates utf8

my proposal isn't to necessarily reuse to_string, but rather, `iobuf_parser` may be a very logical place to put the functionality of reading a string from an iobuf. so feel free to add new and useful utility interfaces like this there.

joe-redpanda replied:
anomaly_detection_test has an odd use case, its sending what is essentially raw binary as an http response because the http mock library requires sstring.
My thought was to avoid bloating iobuf_parser's function list with a utility thats used only once and in a test

further:
I was considering adding a flag to iobuf_parser to allow a read_string to return a shorter string than requested like
```cpp
const auto str = parser.read_string(4096u, allow_insufficient_length::yes);
str.size(); // returns 10
```

the overwhelming majority use case though is either 'read the whole buffer' or getting a specific fixed size string value.

I think it would be better to leaver iobuf_parser simple,
have s3_client manage its own math with a quick std::min,
and have leave anomaly detection as it was before

joe-redpanda replied:
I'm writing the changes real quick

dotnwat replied:
sure. whatever you think is best. just pointing out options that you might not be aware of.

dotnwat commented on src/v/cloud_storage_clients/s3_client.cc:
```diff
@@ -428,9 +430,34 @@ rest_error_response parse_xml_rest_error_response(iobuf&& buf) {
 
+ss::future<rest_error_response> parse_unknown_format_error_response(
+  boost::beast::http::status status, iobuf buf) {
+    static constexpr auto max_body_size = static_cast<size_t>(4096U);
```
```cpp
    #include "base/units.h"
    static constexpr auto max_body_size = 4_KiB;
```

joe-redpanda replied:
I never cease to be amazed by how much syntax candy can be written in vanilla cpp

dotnwat commented on src/v/cloud_storage_clients/s3_client.cc:
```diff
@@ -428,9 +430,34 @@ rest_error_response parse_xml_rest_error_response(iobuf&& buf) {
+
+    co_return rest_error_response{
+      fmt::format("{}", status_to_error_code(status)),
+      fmt::format("http status: {}, error body: {}", status, error_body),
```
you might consider adding the truncation notice here if the log was truncated, instead of using a dedicate INFO level log line for it. something like this

```
error body [truncated~4096]: ...
```

dotnwat commented on src/v/cloud_storage_clients/s3_client.cc:
```diff
@@ -441,18 +468,23 @@ ss::future<ResultT> parse_rest_error_response(
             // https://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html
-            return ss::make_exception_future<ResultT>(
-              parse_xml_rest_error_response(std::move(buf)));
+            // return ss::make_exception_future<ResultT>(
+            //  parse_xml_rest_error_response(std::move(buf)));
```
did you intend to remove these lines?

dotnwat commented on src/v/cloud_storage_clients/s3_error.cc:
```diff
@@ -625,4 +625,15 @@ std::string_view rest_error_response::resource() const noexcept {
+std::ostream& operator<<(std::ostream& o, const rest_error_response err) {
+    static constexpr auto format
+      = "code: {}, message: {}, request_id: {}, resource: {}";
+    return o << fmt::format(
+             format,
```
this builds an intermediate string. you can avoid that by rendering directly to the output stream

```
fmt::print(o, ...);
return o;
```

dotnwat commented:
you need to run the bazel linter

Commit:
cloud_storage_clients/s3_client: Logging Upgrade

s3_client discards the body of error responses if theyre not xml
formatted. This is a problem because some s3 tests periodically fail on
Bad Request with no further description of what failed the request.

This pr adds text logging for non-xml errors.

dotnwat commented on src/v/cloud_storage_clients/s3_client.cc:
```diff
@@ -428,9 +431,38 @@ rest_error_response parse_xml_rest_error_response(iobuf&& buf) {
+    }
+
+    iobuf_parser parser{std::move(buf)};
+    const auto error_body = parser.read_string(read_size);
```
the weird thing here is that utf-8 validation would probably fail and this would throw if the read_size is in the middle of code point. and since we're truncating at an arbitrary point is definitely possible?

mmaslankaprv commented on src/v/cloud_storage_clients/s3_client.cc:
```diff
@@ -428,9 +431,42 @@ rest_error_response parse_xml_rest_error_response(iobuf&& buf) {
 
+rest_error_response parse_unknown_format_error_response(
+  boost::beast::http::status status, iobuf buf) {
+    // static constexpr auto max_body_size = static_cast<size_t>(4_KiB);
+    // static constexpr auto truncation_warning_segment = " [truncated~4096]";
```
Is that to be removed ?

joe-redpanda replied:
Left intentionally. I could bump the TODO comment up

My goal was to 
1. gather s3 error logs with this change (this is because BAD_REQUEST doesn't tell you what went wrong)
2. work on whatever the utf functionality we need to add
3. add truncation back to this with the fixes needed for BAD_REQUEST

The odds a cloud provider sends back a cluster killing response is low, but real enough that truncation is needed eventually

joe-redpanda replied:
This issue doesn't recur on command. The logs were intended to catch the problem when it does eventually recur in CI

dotnwat replied:
> Left intentionally. I could bump the TODO comment up

i think michal might be asking: is this a draft PR or ready for review? if it's ready for review, then it would make sense to remove commented out code unless its part of a bigger comment in which case it would make sense to combine them--otherwise it visually looks like it was forgotten.

joe-redpanda replied:
Ack, got it. Shuffled to formatting to make it clear that this is intentional

dotnwat commented on src/v/cloud_storage_clients/s3_client.cc:
```diff
@@ -428,9 +431,42 @@ rest_error_response parse_xml_rest_error_response(iobuf&& buf) {
+    // }
+
+    iobuf_parser parser{std::move(buf)};
+    const auto error_body = parser.read_string(read_size);
```
won't think still throw a invalid utf8 exception if the read_size happens to land the wrong place?

joe-redpanda replied:
cloud providers promise utf8 & if we're logging the entire error it shouldn't hit any character boundaries

I let this throw because the xml one throws as well.

If we're in the unhappy-unhappy path though, i could see more weird things happening. I'll add a log on utf issue, and then a log + rethrow on anything else

joe-redpanda replied:
We drain the entire response when it arrives. This probably raises a valid question of the value of truncation given that any large response will live in memory at least once

bharathv commented on src/v/cloud_storage_clients/s3_client.cc:
```diff
@@ -428,9 +431,50 @@ rest_error_response parse_xml_rest_error_response(iobuf&& buf) {
+    ss::sstring error_body = "";
+    iobuf_parser parser{std::move(buf)};
+    try {
+        error_body = parser.read_string(read_size);
```
probably a stupid question.. can't we just interpret it as ascii and print 4k to keep it simple? It will probably print all the useful parts of the error? (and good enough in 99% of the cases).

dotnwat replied:
yeh probably. i think there are just three things we want

1. don't accidentally linearize memory that is too big. the input, afaict, is coming from an external system, so we should be resilient.
2. output useful information even if there are problems
3. don't expose harmful output (e.g. half a code point that causes problems for logging system--hopefully they are resilient)

joe-redpanda replied:
This has been a lot of brain cycles on a single log, but the utils that come out I think are worth it

mmaslankaprv replied:
is it safe to read the whole buffer as a string ? I think we should limit the size here

mmaslankaprv replied:
This may lead to a very large memory allocation

joe-redpanda replied:
The truncation is a follow on while I get the truncation functionality through. The odds that we get a crashing response from s3 is low but real. This would at least gather logs from ci for next time we get a BAD_REQUEST and then on merge of the utf utils I would add in truncation

joe-redpanda replied:
incorporated

mmaslankaprv commented on src/v/cloud_storage_clients/s3_client.cc:
```diff
@@ -428,9 +431,50 @@ rest_error_response parse_xml_rest_error_response(iobuf&& buf) {
+    } catch (const invalid_utf8_exception& e) {
+        vlog(s3_log.info, "response body contains invalid utf8 {}", e);
+    } catch (...) {
+        vlog(s3_log.error, "!!error parse error {}", std::current_exception());
```
nit: please avoid `!!` exclamation marks in log messages

joe-redpanda replied:
Can do, can I take this as a fix for the truncation follow on?

joe-redpanda replied:
Per offline discussion we'll block until https://github.com/redpanda-data/redpanda/pull/26697
to make sure truncation is enabled

bharathv replied:
merged https://github.com/redpanda-data/redpanda/pull/26697 FYI.

joe-redpanda replied:
updated

joe-redpanda commented:
/ci-repeat 10
release
skip-redpanda-build
dt-repeat=10
tests/rptest/tests/node_pool_migration_test.py

mmaslankaprv commented on src/v/cloud_storage_clients/s3_client.cc:
```diff
@@ -428,9 +431,48 @@ rest_error_response parse_xml_rest_error_response(iobuf&& buf) {
+
+    // this is the unhappy unhappy path, handle potentially invalid utf8
+    try {
+        error_body = parser.read_string_safe(max_body_size);
```
should that be `read_size` ?

joe-redpanda replied:
Missed, that, ty. Fixed

WillemKauf commented on src/v/cloud_storage_clients/s3_client.cc:
```diff
@@ -428,9 +431,48 @@ rest_error_response parse_xml_rest_error_response(iobuf&& buf) {
+
+    return rest_error_response{
+      fmt::format("{}", status_to_error_code(status)),
+      fmt::format(
```
nit: one less string copy if

```diff
rest_error_response parse_unknown_format_error_response(
  boost::beast::http::status status, iobuf buf) {
    static constexpr auto max_body_size = static_cast<size_t>(4_KiB);
    static constexpr auto truncation_warning_segment = " [truncated~4096]";

-   auto maybe_truncation_warning = "";
+   bool maybe_truncation_warning = false;
    auto read_size = buf.size_bytes();

    // if the error message exceeds reasonable size (4k), truncate it and
    // indicate a truncation to the recipient
    if (read_size > max_body_size) {
-       maybe_truncation_warning = truncation_warning_segment;
+       maybe_truncation_warning = true;
        read_size = max_body_size;
    }

    ss::sstring error_body = "";
    iobuf_parser parser{std::move(buf)};

    // this is the unhappy unhappy path, handle potentially invalid utf8
    try {
        error_body = parser.read_string_safe(read_size);
    } catch (const invalid_utf8_exception& e) {
        vlog(s3_log.info, "response body contains invalid utf8 {}", e);
    } catch (...) {
        vlog(s3_log.error, "error parse error {}", std::current_exception());
        throw;
    }

-    return rest_error_response{
-      fmt::format("{}", status_to_error_code(status)),
-      fmt::format(
-        "http status: {}, error body{}: {}",
-        status,
-        maybe_truncation_warning,
-        error_body),
-      "",
-      ""};
-     return rest_error_response{
+      fmt::format("{}", status_to_error_code(status)),
+     fmt::format(
+      "http status: {}, error body{}: {}",
+        status,
+        maybe_truncation_warning ? truncation_warning_segment : "",
+        error_body),
+      "",
+      ""};
}
```

?

joe-redpanda replied:
maybe_truncation_warning is const char* so theres no extra copy here, just a pointer reassign

WillemKauf replied:
`auto` always throws me for a loop with string literals, fair enough

joe-redpanda replied:
Yeah, this is probably an example of 'auto abuse'. I think in the future I'll explicitly name types when its annoying to parse what auto is doing (e.g. auto blah = std::move(...), is blah a snapped object or a &&?)
